### DISCUSSION TITLE
Best way to handle .env variables

### DISCUSSION THREAD

**Discussion by glenngijsberts** [vercel/next.js#33223] [Help]
Hey there :)

We've recently migrated our app to the new `.env` setup following the documentation on https://nextjs.org/docs/basic-features/environment-variables. However, our app was crashing due to the fact that the `.env` variables were not overwritten by our Kubernetes secrets.

Is it true that the variables from `process.env.*` are always set and build in the code on build-time? How can we determine if the value of a variable needs to be used on runtime? I've seen the documentation of https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration, but does this mean you need to declare variables on 2 places (`next.config.js` and `.env`)?

What are the best practices for handling .env variables both build-time and run-time? How do other people handle this, especially people using Kubernetes secrets as production variables.

**Comment by PsyGik** [comment]
I can't speak for "best practices", (mostly because it's subjective), but I did solve a similar issue in one of the projects I collaborated on which used Next.js. 

We did not use Kubernetes, rather an in-house tool which managed secrets. Here's the gist:

- Create a file that loads secrets from whatever source you got and add secrets to `process.env` 
```
// Call this before starting Next server
const setEnvSecrets = async (secretManagerInstance) => {
  // A glorified promise to resolve secrects  
  const secrets = await secretManagerInstance.getSecrets();
  Object.assign(process.env, ...secrets);
};
```
- Modify your scripts in `package.json`, if using `next start`, to `node secret-loader.js && next start`
- If using a custom server:
```
//index.js
require('./secret-loader');
require('./server');
```
and in `package.json`, `node index.js`
